source/opt/loop_fusion.cpp

```diff
@@ -0,0 +1,735 @@
+                                     instruction, loop_0_);
+                               });
+  induction_1_ = *std::find_if(std::begin(inductions_1), std::end(inductions_1),
+                               [this](ir::Instruction* instruction) {
+                                 return UsedInContinueOrConditionBlock(
+                                     instruction, loop_1_);
+                               });
+
+  if (!CheckInit()) {
+    return false;
+  }
+
+  if (!CheckCondition()) {
+    return false;
+  }
+
+  if (!CheckStep()) {
+    return false;
+  }
+
+  // Check adjacency, |loop_0_| should come just before |loop_1_|.
+  // There is always at least one block between loops, even if it's empty.
+  // We'll check at most 2 preceeding blocks.
+
+  auto pre_header_1 = loop_1_->GetPreHeaderBlock();
+
+  std::vector<ir::BasicBlock*> block_to_check{};
+  block_to_check.push_back(pre_header_1);
+
+  if (loop_0_->GetMergeBlock() != loop_1_->GetPreHeaderBlock()) {
+    // Follow CFG for one more block.
+    auto preds = context_->cfg()->preds(pre_header_1->id());
+    if (preds.size() == 1 && &*containing_function_->FindBlock(preds.front()) ==
+                                 loop_0_->GetMergeBlock()) {
+      block_to_check.push_back(
+          &*containing_function_->FindBlock(preds.front()));
+    } else {
+      return false;
+    }
+  }
+
+  // Check that the separating blocks are either empty or only initialise the
+  // induction variable which isn't used after '--eliminate-local-multi-store'
+  // Also allow OpPhi, since the loop could be in LCSSA form.
+  for (auto block : block_to_check) {
+    for (auto& inst : *block) {
+      if (inst.opcode() == SpvOpStore) {
+        // Get the definition of the target to check it's function scope so
+        // there are no observable side effects.
+        auto variable =
+            context_->get_def_use_mgr()->GetDef(inst.GetSingleWordInOperand(0));
+
+        if (variable->opcode() != SpvOpVariable ||
+            variable->GetSingleWordInOperand(0) != SpvStorageClassFunction) {
+          return false;
+        }
+
+        // Check the target is never loaded.
+        auto is_used = false;
+        context_->get_def_use_mgr()->ForEachUse(
+            inst.GetSingleWordInOperand(0),
+            [&is_used](ir::Instruction* use_inst, uint32_t) {
+              if (use_inst->opcode() == SpvOpLoad) {
+                is_used = true;
+              }
+            });
+
+        if (is_used) {
+          return false;
+        }
+      } else if (inst.opcode() == SpvOpPhi) {
+        if (inst.NumInOperands() != 2) {
+          return false;
+        }
+      } else if (inst.opcode() != SpvOpBranch) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+bool LoopFusion::ContainsOpcode(ir::Loop* loop, SpvOp opcode) {
+  for (const auto& block : loop->GetBlocks()) {
+    for (const auto& inst : *containing_function_->FindBlock(block)) {
```

Comment:
The better way of finding the block is to use `get_inst_block` in the ir context.

Reply:
I'm not sure this is relevant here, as we're not looking for a block that contains a specific instruction.